src/NetMQ/NetMQPoller.cs

```diff
@@ -344,10 +401,13 @@ public void Remove([NotNull] Socket socket)
                 throw new ArgumentNullException(nameof(socket));
             CheckDisposed();
 
-            Run(() =>
+            RunAsync(() =>
             {
-                m_pollinSockets.Remove(socket);
-                m_isPollSetDirty = true;
+                if (m_pollinSockets.Keys.Contains(socket))
+                {
+                    m_pollinSockets.Remove(socket);
+                    m_isPollSetDirty = true;
+                }
```

Comment:
`Dictionary<,>.Remove` returns `bool`, so you can avoid the double-lookup by removing the `Contains` call.

Also, FYI `.Keys.Contains` calls `Enumerable.Contains` which has O(N) performance. If you want to test whether a dictionary contains a key, use `ContainsKey` instead which has O(1) performance.

Reply:
I can make that change, but how many sockets are really going to be in this collection? I doubt enough to matter, nor is this likely to be called often.

Reply:
Ahh... yes, I see.  I was thinking of the indexer `dictionary[key]` .  It is unnecessary.

Reply:
I think I added it trying to track down the error I saw that I ultimately tracked to `NetMQSelector`.

Reply:
removed the check